### PR TITLE
Update manual copies of tape to use tape.copy

### PR DESCRIPTION
**Pull request by astralcai**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/debugging/snapshot.py [hunk 1]
```diff
@@ -189,19 +189,14 @@ def circuit():
...
         else:
             accumulated_ops.append(op)
 
     # Create an additional final tape if a return measurement exists
     if tape.measurements:
         snapshot_tags.append("execution_results")
-        new_tapes.append(
-            type(tape)(ops=accumulated_ops, measurements=tape.measurements, shots=tape.shots)
-        )
+        new_tapes.append(tape.copy(operations=accumulated_ops, measurements=tape.measurements))
```
Comment (mudit2812) [inline pennylane/debugging/snapshot.py:199]:
```suggestion
        new_tapes.append(tape.copy(operations=accumulated_ops))
```

File: pennylane/devices/_qubit_device.py [hunk 2]
```diff
@@ -256,12 +256,7 @@ def execute(self, circuit, **kwargs):
         has_mcm = any(isinstance(op, MidMeasureMP) for op in circuit.operations)
         if has_mcm and "mid_measurements" not in kwargs:
             results = []
-            aux_circ = qml.tape.QuantumScript(
-                circuit.operations,
-                circuit.measurements,
-                shots=[1],
-                trainable_params=circuit.trainable_params,
-            )
+            aux_circ = circuit.copy(shots=[1], trainable_params=circuit.trainable_params)
```
Comment (mudit2812) [inline pennylane/devices/_qubit_device.py:259]:
```suggestion
            aux_circ = circuit.copy(shots=[1])
```

File: pennylane/transforms/decompositions/clifford_t_transform.py [hunk 3]
```diff
@@ -454,7 +454,7 @@ def circuit(x, y):
             decomp_ops.append(qml.GlobalPhase(phase))
 
     # Construct a new tape with the expanded set of operations
-    new_tape = type(tape)(decomp_ops, compiled_tape.measurements, shots=tape.shots)
+    new_tape = compiled_tape.copy(operations=decomp_ops)
```
Comment (mudit2812) [inline pennylane/transforms/decompositions/clifford_t_transform.py:457]:
```suggestion
    new_tape = tape.copy(operations=decomp_ops, measurements=compiled_tape.measurements)
```

Comment (astralcai) [reply pennylane/transforms/decompositions/clifford_t_transform.py:457]: `compiled_tape` comes from `qml.compile`, which has already copied `tape.shots` into `compiled_tape`, so there is no need to be copying the original tape here.

**Review by mudit2812 — APPROVED** [review]

File: pennylane/transforms/optimization/undo_swaps.py [hunk 4]
```diff
@@ -109,8 +109,6 @@ def qfunc():
 
     gates.reverse()
 
-    new_tape = type(tape)(
-        gates, tape.measurements, shots=tape.shots, trainable_params=tape.trainable_params
-    )
+    new_tape = tape.copy(operations=gates)
```
Comment (lillian542) [inline pennylane/transforms/optimization/undo_swaps.py:112]: Was copying the `trainable_params` a mistake here? Changing the operations will set the `trainable_params` back to default, whereas here it looks like we were preserving them explicitly.

**Review by lillian542 — COMMENTED** [review]
Looks good, just one question :)

**Review by lillian542 — APPROVED** [review]
LGTM!

**Pull request merged**